http/src/main/scala/org/http4s/blaze/http/http2/PriorKnowledgeHandshaker.scala

```diff
@@ -0,0 +1,131 @@
+package org.http4s.blaze.http.http2
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+
+import org.http4s.blaze.http.http2.SettingsDecoder.SettingsFrame
+import org.http4s.blaze.pipeline.{Command, TailStage}
+import org.http4s.blaze.util.{BufferTools, Execution}
+
+import scala.concurrent.Future
+
+abstract class PriorKnowledgeHandshaker[T](localSettings: ImmutableHttp2Settings) extends TailStage[ByteBuffer] {
+
+  final protected implicit def ec = Execution.trampoline
+
+  override def name: String = s"${getClass.getSimpleName}($localSettings)"
+
+  /** Handle the prior knowledge preface
+    *
+    * The preface is the magic string "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"
+    * which is intended to cause HTTP/1.x connections to fail gracefully.
+    * For clients, this involves sending the magic string and for servers
+    * this consists of receiving the magic string. The return value of this
+    * function is any unconsumed inbound data.
+    */
+  protected def handlePreface(): Future[ByteBuffer]
+
+  /** Perform pipeline manipulations necessary upon successful handshaking */
+  protected def handshakeComplete(remoteSettings: MutableHttp2Settings, data: ByteBuffer): Future[T]
+
+  /** Perform the HTTP/2 prior knowledge handshake */
+  final def handshake(): Future[T] = {
+    logger.debug("Beginning handshake.")
+    handlePreface()
+      .flatMap(handleSettings)
+      .flatMap { case (settings, acc) =>
+        handshakeComplete(settings, acc)
+      }
+  }
+
+  private[this] def handleSettings(bytes: ByteBuffer): Future[(MutableHttp2Settings, ByteBuffer)] = {
+    sendSettings().flatMap { _ => receiveSettings(bytes) }
+  }
+
+  private[this] def sendSettings(): Future[Unit] = {
+    val settingsBuffer = FrameSerializer.mkSettingsFrame(localSettings.toSeq)
+    channelWrite(settingsBuffer)
+  }
+
+  private[this] def receiveSettings(acc: ByteBuffer): Future[(MutableHttp2Settings, ByteBuffer)] = {
+    logger.debug("receiving settings")
+    getFrameSize(acc) match {
+      case -1 =>
```

Comment:
This is without a settings frame then? Handles getting information till we finally have one? Not sure I follow this section.

Reply:
Correct: if we don't have enough data yet, we read again, concat the data, and try again.

Reply:
I think I wasn't clear: -1 means that there wasn't even enough data for the frame header so a size couldn't be determined. I've changed the result of `getFrameSize` to an `Option[Int]` to better reflect that. Thanks again for pointing that out.